### DISCUSSION TITLE
Advection kernel behaving differently in TET4 mesh compared in HEX8

### DISCUSSION THREAD

**Discussion by schakrabortygithub** [idaholab/moose#24994] [Q&A General]
Hi,

I was trying to use 'ConservativeAdvection' kernel to propagate a scalar field variable along a particular direction (say along x-direction). First I used a HEX8 mesh with 'upwind' correction for the advection term and 'MassLumpedTimeDerivative' for the time derivative term. The model gives expected result. But now if I change the mesh to TET4, I get strange oscillation. Any suggestion to resolve this issue?
I am attaching the field plot and the plot along the z-direction. 

Here is a simple input script to reproduce the problem,
```
[Mesh]
[./gmg]
  type = GeneratedMeshGenerator
  dim = 3
  nx = 10
  ny = 10
  nz = 2
  xmax = 0.1
  ymax = 0.1
  zmax = 0.02
  elem_type = TET4 #TET4, HEX, HEX8
  #show_info = True
[../]
[]

[Outputs]
    file_base = ./Data_CDT_ScalarVar_Temp_TET4
    csv = true
    exodus = true
    #append_date = true
    append_date_format = '%Y-%m-%d-%H-%M-%S'
    execute_on = 'timestep_end'
  [./console]
    type = Console
    output_file = true
    max_rows = 0
  [../]
[]

[Variables]
  [./DD_EdgePositive_00]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[ICs]
  [./IC_DD_EdgePositive_00]
    type = FunctionIC
    variable = DD_EdgePositive_00
    function = '1.0'
  [../]
[]

[Kernels]
  [./dot_DD_EdgePositive_00]
    type = MassLumpedTimeDerivative  #MassLumpedTimeDerivative
    variable = DD_EdgePositive_00
  [../]
  [./advection_DD_EdgePositive_00]
    type = ConservativeAdvection #ConservativeAdvection, AdvectionCDT
    variable = DD_EdgePositive_00
    upwinding_type = Full
    velocity = '0.1 0.0 0.0'
    #block = 0
  [../]

[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON' #'NEWTON' , 'PJFNK'
  #automatic_scaling = true
  #scaling_group_variables = 'disp_x disp_y disp_z; DD_EdgePositive_01'
  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -ksp_type -ksp_gmres_restart'
  petsc_options_value = ' asm      2              lu            gmres     200'
  nl_abs_tol = 1e-5  #1e-10 for all *_tol
  nl_rel_tol = 1e-5
  nl_abs_step_tol = 1e-5

  dt = 0.01
  dtmin = 0.00001
  dtmax = 0.01
  end_time = 1
[]
```

Best
Subhendu

<img width="1228" alt="Screen Shot 2023-07-19 at 11 34 47 PM" src="https://github.com/idaholab/moose/assets/35871741/5b7897e5-6756-49fb-bae8-86bc69c0bb58">

**Comment by GiudGiud** [comment]
Hello

 I am wondering if the mass lumping plays a role here. Can you try with a regular time derivative?

Guillaume

**Reply by schakrabortygithub** [reply]
With regular time derivative, there is no improvement.
Without mass lumping, even HEX8 mesh shows oscillation near boundary or interfaces. As you can see in the attached image, at the right boundary there is a zone with negative gradient of the variable which is unphysical. For longer simulation, it gets worse. Mass lumping eliminates this issue. This is very important because the actual problem that I am trying to solve is to implement '[DiscoFlux](https://doi.org/10.1016/j.ijplas.2015.07.007)' Crystal plasticity model within MOOSE. In this model, dislocation densities are array variable and are transported within each grain using advection model with no flow BC at the interface. The mesh is split with each grain as separate geometry and only displacement is made continuous across interfaces (grain boundary) using penalty method. As the dislocation pile-up at the interface (grain boundary), a separate interface-kernel is used to transport them across the interface from one slip system to the other one based on some geometric criterion. The gradient of the dislocation density (actually the net polarity) contributes towards  the hardening of the material and thus the velocity of dislocations. Hence, capturing the gradient accurately is important. All of these are implemented and worked as expected for simple bi-crystal or rectangular grains with HEX8 mesh but when TET4 mesh is used the above problem arises. And we need to use TET4 mesh for a real polycrystalline material with curved grain boundary.

<img width="1228" alt="Screen Shot 2023-07-20 at 7 57 06 AM" src="https://github.com/idaholab/moose/assets/35871741/9f9afaea-a869-4c21-b8e9-8812be1674c2">
<img width="1228" alt="Screen Shot 2023-07-20 at 8 01 11 AM" src="https://github.com/idaholab/moose/assets/35871741/6c920776-a8e5-4a47-ae3d-4940856e25b5">

**Comment by ABallisat** [comment]
(I might be missing something but...) How do you end up with a value of 3? It looks like your IC is 1 throughout and with a constant velocity that should just shift that constant field sideways, not change the values. You also don't have any BCs which would introduce a change. I've not played with those kernels so I don't know what CFL values it is happy with but seeing as you are at 0.1 for hex elements it might be worth trying with smaller timesteps to see if you get less oscillations. The tets will have smaller side lengths therefore a higher CFL number which may well be less stable.

**Reply by schakrabortygithub** [reply]
Based on my understanding, in the weak form of the advection equation the boundary terms does appear. That can be used to specify the 'inflow' or 'outflow' BC to make the flux go in/out at the boundary. But not specifying the BC is basically 'noflow' BC. The total volume integrated quantity will remain constant and the quantity will get accumulated/depleted at the boundary based on the direction of the advection velocity. The problem that we are trying to solve, we need no-flow BC at the boundary/interface so that the quantity remain preserved. The large value at the boundary than the specified IC is due to this accumulation.

I did try with smaller time-step with fixed dt=0.001, but no improvement. 
Also, I am using implicit time integration. Should be unconditionally stable.
Let me know if you have other thoughts.

Best
Subhendu
Los Alamos National Laboratory

**Comment by lindsayad** [comment]
Does the default nonlinear relative tolerance of 1e-8 not work?

**Reply by schakrabortygithub** [reply]
No, tried with nl_rel_tol=1e-8 and 1e-10 but no benefit.

**Reply by lindsayad** [reply]
@WilkAndy @cpgr do you guys use `ConservativeAdvection`?

**Reply by cpgr** [reply]
I don't sorry!

**Comment by GiudGiud** [comment]
Hello

I took a look.
If you add a Dirichlet BC you can get a good solution (relatively, the left has 0 flux so the solution is probably nonsense still) for TET4 and HEX8.
```
[BCs]
  [inlet]
    type = DirichletBC
    variable = DD_EdgePositive_00
    value = 1
    boundary = right
  []
[]
```

if you don't, it's an advection problem with no boundary value so it does not know what value to get upwind. I think we need a boundary condition to get a well-posed problem.

Guillaume

**Reply by schakrabortygithub** [reply]
Hi Guillaume,

Thanks for the suggestion! 
But specifying the BC still does not solve the problem.
In the attached image (image_1) you can see specifying the Dirichlet BC at the right makes the solution smooth at the surface, as it is expected, but there are oscillation in other parts of the domain. None of these appears in the HEX8 mesh.
Also, if the advection velocity has non zero component in all three direction (image_02) then we need to specify the Dirichlet BC in all boundaries, that will make the problem trivial.
I am wondering if it has to do with not specifying the Dirichlet BC then why the problem does not appear for HEX8 elements, why only for TET4 element?

Also, do you have any suggestion on what kind of BC I should use if I want no flux transfer across the boundary i.e impermeable kind of boundary, and also the total quantity (volume integrated over the domain) to remain constant. Specifying any Dirichlet_BC or flow_BC  at the boundary will not keep the total quantity conserved.

Best
Subhendu


![Picture1](https://github.com/idaholab/moose/assets/35871741/37fa1ea8-b854-451f-a7d5-57e7e30227a6)
![Picture2](https://github.com/idaholab/moose/assets/35871741/d78210fe-a938-448f-be12-b4c52ae1769f)

**Reply by GiudGiud** [reply]
If you have a velocity that s not straight from that face with the dirichlet BC you have an indétermination still. So you need BCs on every face the velocity vector « originated from »

Fig 1 and 2 are good right? Only fig 3 & 4 have issues?
since you have advection in x, y, and z (0.1 in each direction), I d expect a dirichlet BC for 3 faces at lowest x, y and z

**Reply by GiudGiud** [reply]
I reckon you should consider Neumann BCs on the faces. The face value will adapt to get the right advective flux
